server/TransactionService.java

```diff
@@ -132,10 +137,12 @@ public void onError(Throwable error) {
         close(error);
     }
 
-    private synchronized void execute(TransactionProto.Transaction.Req request) {
+    private void execute(TransactionProto.Transaction.Req request) {
```

Comment:
We've replaced our old mutex lock (`synchronized`) with a read-write lock. Read operations acquire a read lock; write operations acquire a write lock.